src/peft/utils/safelora.py

```diff
@@ -0,0 +1,249 @@
+# Copyright 2023-present the HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# Reference paper: https://arxiv.org/abs/2405.16833
+
+
+import copy
+import json
+import os
+from dataclasses import dataclass, field
+
+import numpy
+import torch
+from huggingface_hub import snapshot_download
+from safetensors import SafetensorError, safe_open
+from safetensors.torch import save_file
+from transformers.utils import cached_file
+from transformers.utils.hub import get_checkpoint_shard_files
+
+
+@dataclass
+class SafeLoRAConfig:
+    """
+    This is the configuration class to store the configuration of a safeLoRA.
+    """
+
+    base_model_path: str = field(
+        default=None,
+        metadata={"help": "The path of the base model for obtaining the aligned matrix"},
+    )
+
+    aligned_model_path: str = field(
+        default=None,
+        metadata={"help": "The path of the aligned model for obtaining the aligned matrix"},
+    )
+
+    peft_model_path: str = field(
+        default=None,
+        metadata={"help": "The path of the LoRA wieghts and configs."},
+    )
+
+    select_layers_type: str = field(
+        default="number",
+        metadata={"help": "How to select projection layers? options: [threshold, number]"},
+    )
+
+    threshold: float = field(
+        default=0.5,
+        metadata={"help": "The threshold of cosine similarity."},
+    )
+
+    num_proj_layers: int = field(
+        default=10,
+        metadata={"help": "The number of projected layers."},
+    )
+
+    devices: str = field(
+        default="cuda",
+        metadata={"help": "Devices are used in SafeLoRA. (gpu or cpu)"},
+    )
+
+    saveWeights: bool = field(
+        default=True,
+        metadata={"help": "Replacing and saving SafeLoRA weights to the original LoRA file."},
+    )
+
+    def __post_init__(self):
+        if self.base_model_path is None:
+            raise ValueError("base_model_path cannot be None.")
+        if self.aligned_model_path is None:
+            raise ValueError("aligned_model_path cannot be None.")
+        if self.peft_model_path is None:
+            raise ValueError("peft_model_path cannot be None.")
+
+
+class SafetensorLoader:
```

Comment:
AFAICT, this class is very similar to `loftq_utils._SafetensorLoader`. Instead of copying it with small modifications, would it be possible to re-use the same class? We can move it elsewhere so that it can be more easily imported.

If this proves difficult, please let me know which parts you needed to change and why, and I can help with modifying the class to work in both cases.

Reply:
I want this function to behave like the ```.from_pretrained``` method, where if the user provides a repository name, it will automatically download the files if they are not available locally, and if the user provides a local path, it will load the model directly. However, in the current function, the input needs to be a PEFT model, and it retrieves the base model from the PEFT model's config. Since my input is not a PEFT model, this causes an error.

https://github.com/huggingface/peft/blob/8eeae0a63f28a80ef42b6bfc977c34ffc0b650ff/src/peft/utils/loftq_utils.py#L262C1-L282C27

Reply:
I see. I think this can be fixed. Could you try if this change works:

```diff
modified   src/peft/utils/loftq_utils.py
@@ -269,8 +269,15 @@ class _SafetensorLoader:
 
     def __init__(self, peft_model, model_path):
         if model_path is None:
+            if isinstance(peft_model, str):
+                name_or_path = peft_model
+                base_model_prefix = None
+            else:
+                name_or_path = peft_model
+                base_model_prefix = getattr(peft_model.get_base_model(), "base_model_prefix", None)
+            base_model_prefix = getattr(peft_model.get_base_model(), "base_model_prefix", None)
             try:
-                model_path = snapshot_download(peft_model.base_model.config._name_or_path, local_files_only=True)
+                model_path = snapshot_download(name_or_path, local_files_only=True)
             except (AttributeError, HFValidationError) as exc:
                 raise ValueError(
                     "The provided model does not appear to be a transformers model or is a local model. In this case, "
@@ -286,7 +293,7 @@ class _SafetensorLoader:
             model_path = os.path.join(model_path, suffix)
 
         self.model_path = model_path
-        self.base_model_prefix = getattr(peft_model.get_base_model(), "base_model_prefix", None)
+        self.base_model_prefix = base_model_prefix
         self.prefix = "base_model.model."
         self.is_sharded = False
         self.weight_map = None
```

This way, if the first argument is just a HF model id, it should also work. If this works, we should also rename the arguments, but that can be decided later.

Reply:
I tried the modified version and changed the last line shown below, but it still has issues.

```diff

modified   src/peft/utils/loftq_utils.py
@@ -269,8 +269,15 @@ class _SafetensorLoader:
 
     def __init__(self, peft_model, model_path):
         if model_path is None:
+           if isinstance(peft_model, str):
+                name_or_path = peft_model
+                base_model_prefix = None
+            else:
+                name_or_path = peft_model
+                base_model_prefix = getattr(peft_model.get_base_model(), "base_model_prefix", None)
-            base_model_prefix = getattr(peft_model.get_base_model(), "base_model_prefix", None)
```

1. When the first argument is repo_id (HF model id), the error message is ```LocalEntryNotFoundError: Cannot find an appropriate cached snapshot folder for the specified revision on the local disk and outgoing traffic has been disabled. To enable repo look-ups and downloads online, pass 'local_files_only=False' as input```. To enable automatic model downloading, change ```local_files_only=True``` to ```False``` in ```snapshot_download(name_or_path, local_files_only=True)```.
2. When the first argument is a local path instead of a repo id (HF model id), it cannot pass the section shown below and goes into an exception.  I think it may be necessary to check if the path exists on disk, and only if it doesn’t, proceed with ```snapshot_download```.
```python
    try:
        model_path = snapshot_download(name_or_path,local_files_only=True)
    except (AttributeError, HFValidationError) as exc:
        raise ValueError(
            "The provided model does not appear to be a transformers model or is a local model. In this case, "
            "you must pass the model_path argument that points to the safetensors file."
        ) ```

Reply:
I see, thanks for testing. How about this: We make `local_files_only` an argument to the class. By default, it is `True`, so same as now, but the caller can pass `False`, which would allow the model to be downloaded. Then `apply_safelora` can also take a `local_files_only` argument, which defaults to `False`, and it is passed down to `_SafetensorLoader`. WDYT?

Reply:
I think that sounds feasible and there's no issue with it.

Regarding the two points I mentioned above, the user is inputting a local path instead of a repo_id, so the necessary modifications may be as follows:
```diff

class _SafetensorLoader:
+    def __init__(self, peft_model, model_path=None, local_files_only=True):
        if model_path is None:
            if isinstance(peft_model, str):
                name_or_path = peft_model
                base_model_prefix = None
            else:
                name_or_path = peft_model
                base_model_prefix = getattr(peft_model.get_base_model(), "base_model_prefix", None)
+           if os.path.exists(name_or_path):
+                model_path = name_or_path
+           else:
                try:
+                    model_path = snapshot_download(name_or_path,local_files_only=local_files_only)
                except (AttributeError, HFValidationError) as exc:
                    raise ValueError(
                        "The provided model does not appear to be a transformers model or is a local model. In this case, "
                        "you must pass the model_path argument that points to the safetensors file."
                    ) 

        suffix = "model.safetensors"
        if not model_path.endswith(suffix):
            model_path = os.path.join(model_path, suffix)

```

This version works for both the local path and repo_id. 
If you have any suggestions, please let me know.

Reply:
This sounds good. Let's just rename the `peft_model` argument to `peft_model_or_model_id`. It's a bit ugly but otherwise the name would be confusing.